### PR TITLE
raftexample: fix raft blocking on restart

### PR DESCRIPTION
**Pull request by gyuho**
Separate out 'readSnapshot', replay snapshot in 'newKVStore'

### CODE REVIEW

File: contrib/raftexample/kvstore.go [hunk 1]
```diff
@@ -39,8 +39,15 @@ type kv struct {
 
 func newKVStore(snapshotter *snap.Snapshotter, proposeC chan<- string, commitC <-chan *string, errorC <-chan error) *kvstore {
 	s := &kvstore{proposeC: proposeC, kvStore: make(map[string]string), snapshotter: snapshotter}
-	// replay log into key-value map
-	s.readCommits(commitC, errorC)
+
+	// wait for replayLog signal
```
Comment (heyitsanthony) [inline contrib/raftexample/kvstore.go]: what is `replayLog`?

File: contrib/raftexample/raftexample_test.go [hunk 2]
```diff
@@ -157,3 +160,70 @@ func TestCloseProposerInflight(t *testing.T) {
...
+	old := defaultSnapCount
+	defaultSnapCount = 3
+	defer func() {
+		defaultSnapCount = old
+		os.RemoveAll("raftexample-1")
+		os.RemoveAll("raftexample-1-snap")
+	}()
+
+	startWithSnapshot(t, true)
+	time.Sleep(3 * time.Second)
```
Comment (heyitsanthony) [inline contrib/raftexample/raftexample_test.go]: why 3 seconds?

File: contrib/raftexample/kvstore.go [hunk 3]
```diff
@@ -61,20 +68,24 @@ func (s *kvstore) Propose(k string, v string) {
 	s.proposeC <- string(buf.Bytes())
 }
 
+func (s *kvstore) readSnapshot() error {
+	// replay logs from snapshot
```
Comment (heyitsanthony) [inline contrib/raftexample/kvstore.go]: snapshot loading isn't log replay, though?

**Comment by keekun** [comment]
I encounter the same problem recently. 
Thanks for @gyuho 's tries, I found out a solution and it works.

I think the key of this issue is, `raftexample` try to use a common signal `nil` to represent two different signal: "load snapshot" and "all index applied". Actually they should be treated separately.

In my case, all commit messages are encoded as protobuf, so I just define two string constants, and emit and process them properly.

**Pull request closed without merging**